### ISSUE TITLE
Establishing persistent connection with Spotify App

### ISSUE THREAD

**Issue by udugam** [spotify/ios-sdk#184]
I'm building an app using this SDK and am noticing that my app needs to reconnect with the Spotify app every time my application launches. I understand this is due to the nature of iOS 13, but has anyone figured out a way to establish a persistent connection like the experience that Shazaam offers.

**Comment by ZMcGuckin** [comment]
Did you ever end up figuring this out?

**Comment by kkarayannis** [comment]
Shazam is using an older SDK that is no longer supported. The experience you are currently seeing is the supported one.

**Comment by udugam** [comment]
> Shazam is using an older SDK that is no longer supported. The experience you are currently seeing is the supported one.

**Comment by udugam** [comment]
> Shazam is using an older SDK that is no longer supported. The experience you are currently seeing is the supported one.

Are there plans to mimic the old SDK functionality?

**Comment by udugam** [comment]
> Did you ever end up figuring this out?

Unfortunately we didn’t. Our current user experience with Spotify isn’t as clean and and uninterrupted as our Apple Music integration into the app. Hopefully this can be fixed in the future!

**Comment by ZMcGuckin** [comment]
@udugam Hey, my team ended up getting it work fairly well lately, and it's about 95% as good as the Apple Music integration on iOS now. We found using the workaround they have for syncing without playing music (at the bottom of the README), and using our own system we had in replace for token refreshing allowed us to connect and stay connected well. Even with disconnecting and reconnecting upon entering the background as they suggest.

**Comment by anscash** [comment]
@ZMcGuckin 
Sounds VERY interesting. Thanks for sharing your information. I'm also very interested in this subject.
What README to you mean? The Spotify README or the Apple Music README?

**Comment by ZMcGuckin** [comment]
@anscash The Spotify SDK README

**Comment by anscash** [comment]
Thank you very much. 
So you are doing this successfully? Are you now able to auth with Spoitfy without jumping to the Spotify app every time you launch your app?
What if you're trying to play a track? Do you have to perform the app-switch then?

I am asking kind of stupid because i can't believe it :-D

**Comment by ZMcGuckin** [comment]
Yes! It's all working with minimal to no Spotify switching. The only switch I do is to either sync it because they're not currently using Spotify, which will pull up their most recent track in Spotify, and then when they're playing music already and open the app the sdk automatically connects and shows the connection without switching

**Comment by anscash** [comment]
Thanks, sounds very interesing! :-)
But if my user does not play any spotify song in the moment (Spotify app = closed/inactive), my app has still redirect to Spotify to establish the connection, right?

**Comment by ZMcGuckin** [comment]
Yes, you would need to build the "tunnel" to Spotify in that case, can't be avoided

**Comment by anscash** [comment]
Okay, thank you :-)

**Comment by njsa04** [comment]
@ZMcGuckin Interesting what you mention about using your own server for token refreshing. 
In short, are you saying you would have your app activate spotify the first time via initiateSessionWithScope to get the initial valid token and after that point using session.expired to determine if the token had expired and if so, you would pass the invalid token to your own swap endpoint to renew and then manually setting appRemote.connectionParams.accessToken = yourManualServerRenewedToken to then continue playing and controlling the sdk via appremote?

**Comment by ZMcGuckin** [comment]
@njsa04 yeah exactly. I store in my own database the refresh time and when it was last updated and if it’s expired I contact the Spotify api to refresh it and use the new valid access token

**Comment by njsa04** [comment]
Thanks, curious if you've also dealt with the issue where Spotify had been in background playing audio or perhaps not playing any audio for several hours let's say, and your app launches it to open the tunnel but app remote ceases to connect and only way to recover is force quit spotify app and let your app open it to launch a fresh connection?

**Comment by ZMcGuckin** [comment]
@njsa04  yeah that’s still an issue Spotify needs to fix

**Comment by njsa04** [comment]
Thanks for confirming above, really makes the iOS / Spotify experience clunky.

Have you found a way to detect that situation programmatically so that the user can be notified to close Spotify and retry?

**Comment by ZMcGuckin** [comment]
Unfortunately no, I do feel like it would be easy to find out programmatically, but requesting the user to close the Spotify app seems a bit odd so I haven’t
